perma_web/api2/serializers.py

```diff
@@ -0,0 +1,242 @@
+from django.conf import settings
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.core.validators import URLValidator
+from requests import TooManyRedirects
+from rest_framework import serializers
+
+from utils import get_mime_type, mime_type_lookup
+from perma.models import LinkUser, Folder, CaptureJob, Capture, Link, Organization
+from perma.utils import ip_in_allowed_ip_range
+
+
+class BaseSerializer(serializers.ModelSerializer):
+    """ Base serializer from which all of our serializers inherit. """
+
+    def update(self, instance, validated_data):
+        """ When updating, requiring that our serializers provide a whitelist of fields that can be updated. """
+        if set(validated_data.keys()) - set(self.Meta.allowed_update_fields):
+            raise serializers.ValidationError('Only updates on these fields are allowed: %s' % ', '.join(self.Meta.allowed_update_fields))
+        return super(BaseSerializer, self).update(instance, validated_data)
+
+
+### LINKUSER ###
+
+class LinkUserSerializer(BaseSerializer):
+    full_name = serializers.ReadOnlyField(source='get_full_name')
+    short_name = serializers.ReadOnlyField(source='get_short_name')
+    top_level_folders = serializers.SerializerMethodField()
+
+    class Meta:
+        model = LinkUser
+        fields = ('id', 'first_name', 'last_name', 'full_name', 'short_name', 'top_level_folders')
+
+    def get_field_names(self, *args, **kwargs):
+        # Exclude the top_level_folders field when this serializer is nested inside another serializer.
+        # This lets us use the same serializer at /user and also to embed created_by user info in /archives
+        field_names = super(LinkUserSerializer, self).get_field_names(*args, **kwargs)
+        if self.parent:
+            field_names = [i for i in field_names if i != 'top_level_folders']
+        return field_names
+
+    def get_top_level_folders(self, user):
+        serializer = FolderSerializer(user.top_level_folders(), many=True)
+        return serializer.data
+
+
+### FOLDER ###
+
+class FolderSerializer(BaseSerializer):
+    has_children = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Folder
+        fields = ('id', 'name', 'parent', 'has_children', 'organization')
+        extra_kwargs = {'parent': {'required': True}}
+        allowed_update_fields = ['name', 'parent']
+
+    def get_has_children(self, folder):
+        return not folder.is_leaf_node()
+
+    def validate_name(self, name):
+        if self.instance:
```

Comment:
What does it mean if self.instance is falsy here? Does that mean we are in the process of trying to create a new folder, and so don't yet have a saved folder object? (I do see that the DRF has a default value of None for self.instance)